Fix linter after Go 1.18 upgrade

zmb3 opened the pull request:
We need to pull in a newer version of golangci-lint which vendors linters that are compatible with 1.18.

Additionally, fix new warnings that came up as a result of the ugprade.

Lastly, disable the typecheck linter because it fails due to our use of the `e` submodule. Fear not, compile errors will still be caught by the CI build.

zmb3 committed:
Update golangci-lint

To accomodate the recent Go 1.18 upgrade

zmb3 committed:
Fix new lint warnings as a result of linter upgrade

espadolini reviewed (approved):
I'm not 100% sure about the use of `(net.Error).Timeout()` as a substitute for `.Temporary()` (and neither is the go http package apparently) but otherwise LGTM

zmb3 committed:
Set golangci-lint to Go 1.18 mode

golangci-lint will automatically skip linters that don't have support
for Go 1.18.

See: https://github.com/golangci/golangci-lint/issues/2649

espadolini commented:
Can't we just wait until we actually bump the version in go.mod?

zmb3 commented:
> Can't we just wait until we actually bump the version in go.mod? (re: [4c07973](https://github.com/gravitational/teleport/commit/4c07973ed548caf8f49483b49cf36b6cc338f97c))

We could, but this is semantically more correct, as the Go directive in go.mod specifies the _minimum Go version needed to compile the module_. We aren't using any Go 1.18 features yet, and things happily compile with 1.17, so I felt like this was not yet warranted.

zmb3 commented:
I'm not sure why `gocritic` is even trying to run, it's disabled by default and not enabled in our config. In fact, when I run locally I see as much:

```
golangci-lint run -c .golangci.yml --build-tags='' 
WARN [linters context] bodyclose is disabled because of go1.18. You can track the evolution of the go1.18 support by following the https://github.com/golangci/golangci-lint/issues/2649. 
WARN [linters context] gosimple is disabled because of go1.18. You can track the evolution of the go1.18 support by following the https://github.com/golangci/golangci-lint/issues/2649. 
WARN [linters context] staticcheck is disabled because of go1.18. You can track the evolution of the go1.18 support by following the https://github.com/golangci/golangci-lint/issues/2649. 
WARN [linters context] structcheck is disabled because of go1.18. You can track the evolution of the go1.18 support by following the https://github.com/golangci/golangci-lint/issues/2649. 
WARN [linters context] unused is disabled because of go1.18. You can track the evolution of the go1.18 support by following the https://github.com/golangci/golangci-lint/issues/2649.
```

codingllama commented:
Disable gocritic entirely, so you have peace of mind to try and fix it after this lands?

zmb3 commented:
`gocritic` is not enabled. That's the strange thing.

Our config uses `disable-all: true` and then enables specific linters, gocritic is not one of them.

zmb3 commented:
Ah, we have a chicken and egg problem. The build won't run with the new image until this PR is merged, and the PR can't be merged until the linter is fixed. /sigh

espadolini commented:
Is there any fix other than turning the lint rule into a noop for one commit?

zmb3 commented:
Yes, we're manually updating the buildbox image now.
